lib/IDE/CodeCompletionResultType.cpp

```diff
@@ -11,50 +11,261 @@
+  // the same USR.
+  Ty = Ty->getCanonicalType();
+
+  // For opaque types like 'some View', consider them equivalent to 'View'.
+  if (auto OpaqueType = Ty->getAs<ArchetypeType>()) {
+    if (auto Existential = OpaqueType->getExistentialType()) {
+      Ty = Existential;
+    }
+  }
+  // We can't represent more complicated archetypes like 'some View & MyProto'
+  // in USRBasedType yet. Simply map them to null types for now.
+  if (Ty->hasArchetype()) {
+    return USRBasedType::null(Arena);
+  }
+
+  SmallVector<const USRBasedType *, 2> Supertypes;
+  if (auto Nominal = Ty->getAnyNominal()) {
+    // Sorted conformances so we get a deterministic supertype order and can
+    // assert that USRBasedTypes with the same USR have the same supertypes.
+    auto Conformances = Nominal->getAllConformances(/*sorted=*/true);
+    Supertypes.reserve(Conformances.size());
+    for (auto Conformance : Conformances) {
+      if (Conformance->getDeclContext()->getParentModule() !=
+          Nominal->getModuleContext()) {
+        // Only include conformances that are declared within the module of the
+        // type to avoid caching retroactive conformances which might not
+        // exist when using the code completion cache from a different module.
+        continue;
+      }
+      Supertypes.push_back(USRBasedType::fromType(
+          Conformance->getProtocol()->getDeclaredInterfaceType(), Arena));
+    }
+  }
+  Type Superclass = Ty->getSuperclass();
+  while (Superclass) {
+    Supertypes.push_back(USRBasedType::fromType(Superclass, Arena));
+    Superclass = Superclass->getSuperclass();
+  }
+
+  SmallString<32> USR;
+  llvm::raw_svector_ostream OS(USR);
+  printTypeUSR(Ty, OS);
+
+  assert(llvm::all_of(Supertypes, [&USR](const USRBasedType *Ty) {
+    return Ty->getUSR() != USR;
+  }) && "Circular supertypes?");
+
+  llvm::SmallPtrSet<const USRBasedType *, 2> ImpliedSupertypes;
+  for (auto Supertype : Supertypes) {
+    ImpliedSupertypes.insert(Supertype->getSupertypes().begin(),
+                             Supertype->getSupertypes().end());
+  }
+  llvm::erase_if(Supertypes, [&ImpliedSupertypes](const USRBasedType *Ty) {
+    return ImpliedSupertypes.contains(Ty);
+  });
+
+  return USRBasedType::fromUSR(USR, Supertypes, Arena);
+}
+
+TypeRelation USRBasedType::typeRelation(const USRBasedType *ResultType,
+                                        const USRBasedType *VoidType) const {
+  // `this` is the contextual type.
+  if (this == VoidType) {
+    // We don't report Void <-> Void matches because that would boost
+    // methods returning Void in e.g.
+    // func foo() { #^COMPLETE^# }
+    // because #^COMPLETE^# is implicitly returned. But that's not very
+    // helpful.
+    return TypeRelation::Unknown;
+  }
+  if (ResultType == this) {
+    return TypeRelation::Identical;
+  }
+  for (auto *Supertype : ResultType->getSupertypes()) {
+    if (this->typeRelation(Supertype, VoidType) >= TypeRelation::Convertible) {
```

Comment:
When the hierarchy contains diamond conformance,
```swift
protocol Base {}
protocol P1: Base {}
protocol P2: Base {}
struct S: P1, P2 {}
```
I think `Base` is visited twice. I'm not sure this happens lot, and not sure it causes excessive waste. But this logic should have "visited" check.

Reply:
Added it https://github.com/apple/swift/pull/41000/commits/b12d82a3e811a22aedadb06a9a46af20517be664